partial fix for Issue 12760 Title gets updated after profile modal closes

aevix opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

Fixes #12760 


### Changes
@Alys It is still throwing some errors but its doing what we are wanting to do so I will link this PR. Its a bit of a long explanation sorry... So I finally understand what @paglias meant when he said this is caused by router not being standard. Modals are handled through bootstrap vue and so when it opens the modal the whole app doesn't need to re render since nothing changed in the app. So I tried to create a new router for the profile modal thinking its just another component that needs to be hooked in but that got very messy and not exactly doing what I thought it would. So I went back to see how the profile component was being routed (not the modal) and tried just mimicking what it was doing but with no luck, it wasn't reacting at all until I was about to give up and tried adding this.$router.push() instead of .go(-1) and it finally updated the title! I am not sure 100% but some errors are being thrown and it may have something to do with promise with routing but if someone knows what is going on feel free to let me know. If someone has better idea or want me to change anything let me know also.



----
UUID: 228f739f-b42c-4801-932c-10b4d739b460

aevix committed:
issue-12760

title change after profilemodal being opened

aevix committed:
Update profileModal.vue

aevix committed:
Update profileModal.vue

This commit will change the title to the correct string even when changing to different paths like time travlers, group, etc.

paglias commented on website/client/src/components/userMenu/profileModal.vue:
```diff
@@ -54,8 +61,40 @@ export default {
     onHidden () {
       if (this.$route.path !== window.location.pathname) {
-        this.$router.go(-1);
+        this.$router.push({ path: this.$route.path });
```
Hi @aevix , I'm seeing a few problems with this, instead of trying to recreate the previous title , my suggestion would be to store it in a variable before it's set [here](https://github.com/HabitRPG/habitica/blob/316ed391e8c92894bdafe31a41bb8ec6d1e3f418/website/client/src/components/userMenu/profile.vue#L902) and then restore it when `onHidden` is called. I made an example here https://github.com/HabitRPG/habitica/issues/12760#issuecomment-725464599

aevix replied:
I am not sure if I am not doing something wrong but I have tried this already and it won't change the title. Maybe you could try it and see if it works on your end. Maybe its the browser I am using but I tested it on safari and google chrome. document.title from the app.js is being triggered and working via `common:setTitle` but I think like I mentioned in the issue, vue router has some sort of disconnect between dom being rendered and the actual browser updating. Only way I got it to update was by changing it to push.

paglias replied:
could you post the commit where you tried that and I'll give it a look? thanks

aevix replied:
I made a commit sorry it took so long to respond I didn't get a chance to sit down and work on it until now.

paglias commented on website/client/src/components/userMenu/profileModal.vue:
```diff
@@ -35,15 +35,22 @@ export default {
       this.path = data.path;
       this.$root.$emit('bv::show::modal', 'profile');
     });
+    this.selectPage(this.startingPage);
```
why is this necessary?

paglias replied:
Actually I'm not seeing where this function is declare or `this.selectedPage` used

aevix replied:
Oh I must have left it by accident I was testing some code and forgot to take it out. Ill push a commit without that function.

gryphonbot committed:
Translated using Weblate (Spanish (Latin America))

Currently translated at 87.5% (1920 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/es_419/

Translated using Weblate (Italian)

Currently translated at 100.0% (718 of 718 strings)

Translation: Habitica/Questscontent
Translate-URL: https://translate.habitica.com/projects/habitica/questscontent/it/

Translated using Weblate (Italian)

Currently translated at 100.0% (363 of 363 strings)

Translation: Habitica/Content
Translate-URL: https://translate.habitica.com/projects/habitica/content/it/

Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (23 of 23 strings)

Translation: Habitica/Loginincentives
Translate-URL: https://translate.habitica.com/projects/habitica/loginincentives/pt_BR/

Translated using Weblate (Japanese)

Currently translated at 91.4% (2006 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/ja/

Translated using Weblate (Czech)

Currently translated at 87.5% (1921 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/cs/

Translated using Weblate (Czech)

Currently translated at 87.5% (1921 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/cs/

Translated using Weblate (Norwegian Bokmål)

Currently translated at 100.0% (111 of 111 strings)

Translation: Habitica/Pets
Translate-URL: https://translate.habitica.com/projects/habitica/pets/nb_NO/

Translated using Weblate (Hindi)

Currently translated at 46.8% (52 of 111 strings)

Translation: Habitica/Pets
Translate-URL: https://translate.habitica.com/projects/habitica/pets/hi/

Translated using Weblate (Latin)

Currently translated at 98.9% (185 of 187 strings)

Translation: Habitica/Character
Translate-URL: https://translate.habitica.com/projects/habitica/character/la/

Translated using Weblate (Norwegian Bokmål)

Currently translated at 100.0% (134 of 134 strings)

Translation: Habitica/Tasks
Translate-URL: https://translate.habitica.com/projects/habitica/tasks/nb_NO/

Translated using Weblate (Irish)

Currently translated at 84.6% (11 of 13 strings)

Translation: Habitica/Rebirth
Translate-URL: https://translate.habitica.com/projects/habitica/rebirth/ga/

Translated using Weblate (Norwegian Bokmål)

Currently translated at 100.0% (8 of 8 strings)

Translation: Habitica/Overview
Translate-URL: https://translate.habitica.com/projects/habitica/overview/nb_NO/

Translated using Weblate (Hindi)

Currently translated at 91.8% (90 of 98 strings)

Translation: Habitica/Achievements
Translate-URL: https://translate.habitica.com/projects/habitica/achievements/hi/

Translated using Weblate (Norwegian Bokmål)

Currently translated at 90.1% (55 of 61 strings)

Translation: Habitica/Messages
Translate-URL: https://translate.habitica.com/projects/habitica/messages/nb_NO/

Translated using Weblate (Hindi)

Currently translated at 96.0% (195 of 203 strings)

Translation: Habitica/Generic
Translate-URL: https://translate.habitica.com/projects/habitica/generic/hi/

Translated using Weblate (Norwegian Bokmål)

Currently translated at 77.6% (1703 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/nb_NO/

Translated using Weblate (Irish)

Currently translated at 77.5% (1701 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/ga/

Translated using Weblate (Norwegian Bokmål)

Currently translated at 92.4% (196 of 212 strings)

Translation: Habitica/Settings
Translate-URL: https://translate.habitica.com/projects/habitica/settings/nb_NO/

Translated using Weblate (Spanish (Latin America))

Currently translated at 86.9% (1907 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/es_419/

Translated using Weblate (Dutch)

Currently translated at 97.6% (2141 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/nl/

Translated using Weblate (Spanish)

Currently translated at 87.4% (1918 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/es/

Translated using Weblate (Czech)

Currently translated at 87.5% (1921 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/cs/

Translated using Weblate (Romanian)

Currently translated at 100.0% (187 of 187 strings)

Translation: Habitica/Character
Translate-URL: https://translate.habitica.com/projects/habitica/character/ro/

Translated using Weblate (Bulgarian)

Currently translated at 100.0% (187 of 187 strings)

Translation: Habitica/Character
Translate-URL: https://translate.habitica.com/projects/habitica/character/bg/

Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (2193 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/pt_BR/

Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (23 of 23 strings)

Translation: Habitica/Loginincentives
Translate-URL: https://translate.habitica.com/projects/habitica/loginincentives/pt_BR/

Translated using Weblate (Japanese)

Currently translated at 91.2% (2002 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/ja/

Translated using Weblate (Spanish)

Currently translated at 87.4% (1918 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/es/

Translated using Weblate (Japanese)

Currently translated at 91.1% (1998 of 2193 strings)

Translation: Habitica/Gear
Translate-URL: https://translate.habitica.com/projects/habitica/gear/ja/

Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (15 of 15 strings)

Translation: Habitica/Death
Translate-URL: https://translate.habitica.com/projects/habitica/death/pt_BR/

Translated using Weblate (French)

Currently translated at 100.0% (98 of 98 strings)

Translation: Habitica/Achievements
Translate-URL: https://translate.habitica.com/projects/habitica/achievements/fr/

Translated using Weblate (French)

Currently translated at 100.0% (98 of 98 strings)

Translation: Habitica/Achievements
Translate-URL: https://translate.habitica.com/projects/habitica/achievements/fr/

aevix committed:
Update profileModal.vue

deleted unecessary function call in the mounted hook.

gryphonbot committed:
Merge branch 'origin/develop' into Weblate.

aevix committed:
Set page title using profile

on profile I set the old title before modal changes the title and then when onhidden is triggered I fired off an emit and have profile catch the emit and change the title back to the old title using $store.dispatch.

thewilloftheshadow committed:
make guilds not display group plans as title section (#12761)

* make guilds not display group plans as title section

* fix for parties as well

* task board for group plans now says group plans

* fix party title

Co-authored-by: Matteo Pagliazzi <matteopagliazzi@gmail.com>

paglias commented on website/client/src/components/userMenu/profileModal.vue:
```diff
@@ -35,6 +35,7 @@ export default {
       userId: undefined,
       startingPage: undefined,
       path: undefined,
+      selectedPage: '',
```
this can be removed I think

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -838,6 +839,11 @@ export default {
   mounted () {
     this.loadUser();
     this.selectPage(this.startingPage);
+    this.$root.$on('restoreTitle', () => {
```
to make sure we don't keep this event listener even after the component is not being used it should be removed before the component is destroyed with the following hook(you can place it right after the `mounted` hook

```
  beforeDestroy () {
    this.$root.$off('habitica: restoreTitle');
  },
```

paglias commented on website/client/src/components/userMenu/profile.vue:
```diff
@@ -899,6 +905,7 @@ export default {
     selectPage (page) {
       this.selectedPage = page || 'profile';
       window.history.replaceState(null, null, '');
+      this.oldTitle = this.$store.state.title;
```
I haven't tested this @aevix but my guess is that `selectPage` is called not only the first time the modal is opened but also when switching, for example, from Stats to Achievements, in which case it would store the wrong title. I think this should be called in `mounted` so it's called just once

paglias commented on website/client/src/components/userMenu/profileModal.vue:
```diff
@@ -55,6 +56,7 @@ export default {
     onHidden () {
       if (this.$route.path !== window.location.pathname) {
         this.$router.go(-1);
+        this.$root.$emit('restoreTitle');
```
to follow internal conventions, can you rename the event `habitica: restoreTitle`?

paglias reviewed (changes requested):
@aevix left some minor comments but it's looking good! It's a very clean and concise solution

paglias commented:
Also there seems to be an issue with the branch, it should target `develop` not `delta` (but if it's an issue don't worry I'll fix it before merging)

aevix commented:
@paglias This commit doesn't solve the issue. I am not sure if you have tried running this but on my browser title doesn't get updated like it should. I just created the commit to show that this way doesn't seem to work I am not sure why it won't update. My previous commit f35030c does update the title but it throws errors on the console. Do you still want me to edit the commit with your comments?

aevix committed:
update profile and profileModal from comments

I made the changes with the comments but this still doesn't seem to work. If this is not just a issue on my end I still think its some sort of rendering issue with the browser and router. I checked to see if document.title does get updated after the modal gets hidden and it does update the title but browser doesn't change.

paglias commented:
@aevix I got confused and hadn't tested it earlier, I tested it today and with a few changes it's going to work!

In particular:

- The old title was being saved after the new was already set by `this.startingPage()`
- The event was being fired after the `hidden` event which meant the profile component had already been updated.

This is the diff of the changes I applied (I also changed the event name to `habitica:restoreTitle` without the whitespace):

```

paglias commented:
@aevix I got confused and hadn't tested it earlier, I tested it today and with a few changes it's going to work!

In particular:

- The old title was being saved after the new was already set by `this.startingPage()`
- The event was being fired after the `hidden` event which meant the profile component had already been updated.

This is the diff of the changes I applied to make it work locally (I also changed the event name to `habitica:restoreTitle` without the whitespace):

```
matteo@macbook-matteo habitica % git diff
diff --git a/website/client/src/components/userMenu/profile.vue b/website/client/src/components/userMenu/profile.vue
index 01dfd1e498..bb33920c1c 100644
--- a/website/client/src/components/userMenu/profile.vue
+++ b/website/client/src/components/userMenu/profile.vue
@@ -782,7 +782,7 @@ export default {
       content: Content,
       user: null,
       userLoaded: false,
-      oldTitle: '',
+      oldTitle: null,
     };
   },
   computed: {
@@ -838,16 +838,18 @@ export default {
   },
   mounted () {
     this.loadUser();
-    this.selectPage(this.startingPage);
     this.oldTitle = this.$store.state.title;
-    this.$root.$on('habitica: restoreTitle', () => {
-      this.$store.dispatch('common:setTitle', {
-        fullTitle: this.oldTitle,
-      });
+    this.selectPage(this.startingPage);
+    this.$root.$on('habitica:restoreTitle', () => {
+      if (this.oldTitle) {
+        this.$store.dispatch('common:setTitle', {
+          fullTitle: this.oldTitle,
+        });
+      }
     });
   },
   beforeDestroy () {
-    this.$root.$off('habitica: restoreTitle');
+    this.$root.$off('habitica:restoreTitle');
   },
   methods: {
     async loadUser () {
diff --git a/website/client/src/components/userMenu/profileModal.vue b/website/client/src/components/userMenu/profileModal.vue
index 251cb3ba75..06b2611cc1 100644
--- a/website/client/src/components/userMenu/profileModal.vue
+++ b/website/client/src/components/userMenu/profileModal.vue
@@ -4,6 +4,7 @@
     size="lg"
     :hide-footer="true"
     :hide-header="true"
+    @hide="beforeHide"
     @hidden="onHidden"
     @shown="onShown()"
   >
@@ -52,10 +53,14 @@ export default {
     onShown () {
       window.history.pushState('', null, this.path);
     },
+    beforeHide () {
+      if (this.$route.path !== window.location.pathname) {
+        this.$root.$emit('habitica:restoreTitle');
+      }
+    },
     onHidden () {
       if (this.$route.path !== window.location.pathname) {
         this.$router.go(-1);
-        this.$root.$emit('habitica: restoreTitle');
       }
     },
   },
(END)
```

paglias committed:
fix issue, reset to develop

paglias committed:
Merge branch 'issue-12760' of https://github.com/aevix/habitica into issue-12760

paglias commented:
Unfortunately fixing the PR base was not possible (you probably created your branch from the `delta` one instead of develop originally) so I merged it manually https://github.com/HabitRPG/habitica/commit/df1e4af7fc244e929c900a842048cf39f8dead61 with a direct commit.

@aevix thanks for this PR! I'm awarding your contributor tier.

Note that further tiers require increasing amounts of work from one to the next, so Tier 2 may require a couple of PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!

aevix commented:
@paglias Thank you so much for showing me what you did. I was completely stumped on this issue and probably bit off a little more than I could chew lol. I will continue to help out any way I can 😄
